### PR TITLE
migrant codeblocks validation from apex to docs

### PR DESCRIPTION
**Pull request by xichan-ms**
###### Microsoft Reviewers: [Open in CodeFlow](http://wpcp.azurewebsites.net/CodeFlowProtocolProxy2.php?pullrequest=https://github.com/dotnet/docfx/pull/6356)

### CODE REVIEW

**Comment by dnfadmin** [comment]
[CLA assistant check](https://cla.dotnetfoundation.org/dotnet/docfx?pullRequest=6356) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.dotnetfoundation.org/dotnet/docfx?pullRequest=6356) before we can accept your contribution.<br/><br/>:x: xichan-ms [sign now](https://cla.dotnetfoundation.org/dotnet/docfx?pullRequest=6356)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.dotnetfoundation.org/check/dotnet/docfx?pullRequest=6356) it.</sub>

File: src/docfx/validation/ContentValidator.cs [hunk 1]
```diff
@@ -55,6 +55,15 @@ public void ValidateImageLink(Document file, SourceInfo<string> link, string? al
             }
         }
 
+        public void ValidateCodeBlock(Document file, CodeBlockItem codeBlockItem, bool v)
```
Comment (hejingkan2005) [inline src/docfx/validation/ContentValidator.cs:58]: Remove "bool v"?

Comment (xichan-ms) [reply src/docfx/validation/ContentValidator.cs:58]: OK！

File: src/docfx/lib/markdown/MarkdigUtility.cs [hunk 2]
```diff
@@ -180,9 +182,21 @@ public static void Visit(this MarkdownObject? obj, MarkdownVisitContext context,
...
+                        }
+                        context.TripleColonCount--;
+                    }
+                    else
+                    {
+                        foreach (var child in block)
+                        {
+                            Visit(child, context, action);
+                        }
                     }
```
Comment (OsmondJiang) [inline src/docfx/lib/markdown/MarkdigUtility.cs:200]: doesn't make sense

File: src/docfx/lib/markdown/validation/DocsValidationExtension.cs [hunk 3]
```diff
@@ -80,6 +82,32 @@ internal static class DocsValidationExtension
                         }
                     }
 
+                    CodeBlockItem? codeBlockItem = null;
+                    if (node.GetType().Name.Equals(nameof(CodeBlock)) && context.TripleColonCount == 0)
```
Comment (OsmondJiang) [inline src/docfx/lib/markdown/validation/DocsValidationExtension.cs:86]: >node.GetType().Name.Equals(nameof(CodeBlock)) [](start = 24, length = 45)

node is CodeBlock

File: src/docfx/lib/markdown/MarkdownVisitContext.cs [hunk 4]
```diff
@@ -16,6 +16,8 @@ internal class MarkdownVisitContext
 
         public IEnumerable<SourceInfo?> Parents => FileStack.Reverse().Skip(1).Select(f => f.Source);
 
+        public int TripleColonCount { get; set; } = 0;
```
Comment (OsmondJiang) [inline src/docfx/lib/markdown/MarkdownVisitContext.cs:19]: >TripleColonCount [](start = 19, length = 16)

Stack<ITripleColonExtensionnInfo>

File: src/docfx/lib/markdown/validation/DocsValidationExtension.cs [hunk 5]
```diff
@@ -80,6 +82,32 @@ internal static class DocsValidationExtension
...
+                            Arguments = fencedCodeBlock.Arguments,
+                            IsOpen = fencedCodeBlock.IsOpen,
+                            SourceInfo = fencedCodeBlock.GetSourceInfo(),
+                        };
+                    }
+
+                    if (codeBlockItem != null)
+                    {
+                        codeBlockItemList.Add(codeBlockItem);
+                    }
```
Comment (OsmondJiang) [inline src/docfx/lib/markdown/validation/DocsValidationExtension.cs:109]: extract a method for code block validation

File: src/docfx/lib/markdown/validation/DocsValidationExtension.cs [hunk 6]
```diff
@@ -80,6 +82,32 @@ internal static class DocsValidationExtension
                         }
                     }
```
Comment (OsmondJiang) [inline src/docfx/lib/markdown/validation/DocsValidationExtension.cs:83]: extract a method for heading validation

Comment (xichan-ms) [reply src/docfx/lib/markdown/validation/DocsValidationExtension.cs:83]: ok

File: src/docfx/lib/markdown/validation/DocsValidationExtension.cs (hunk 5 above)
Comment (xichan-ms) [reply src/docfx/lib/markdown/validation/DocsValidationExtension.cs:109]: ok

File: src/docfx/lib/markdown/MarkdownVisitContext.cs (hunk 4 above)
Comment (xichan-ms) [reply src/docfx/lib/markdown/MarkdownVisitContext.cs:19]: ok

File: src/docfx/lib/markdown/validation/DocsValidationExtension.cs (hunk 3 above)
Comment (xichan-ms) [reply src/docfx/lib/markdown/validation/DocsValidationExtension.cs:86]: ok

File: src/docfx/lib/markdown/MarkdigUtility.cs (hunk 2 above)
Comment (xichan-ms) [reply src/docfx/lib/markdown/MarkdigUtility.cs:200]: ok

File: src/docfx/lib/markdown/validation/DocsValidationExtension.cs [hunk 7]
```diff
@@ -88,9 +50,108 @@ internal static class DocsValidationExtension
...
+                {
+                    Type = CodeBlockTypeEnum.FencedCodeBlock,
+                    Info = fencedCodeBlock.Info,
+                    Arguments = fencedCodeBlock.Arguments,
+                    IsOpen = fencedCodeBlock.IsOpen,
+                    SourceInfo = fencedCodeBlock.GetSourceInfo(),
+                    IsCanonicalVersion = isCanonicalVersion,
+                };
+            }
+            else if (node is CodeBlock codeBlock && context.TripleColonStack.Count == 0)
```
Comment (yufeih) [inline src/docfx/lib/markdown/validation/DocsValidationExtension.cs:139]: Are there more context around this check `context.TripleColonStack.Count == 0`?

Triple colon is a broad concept, it is a general way of doing any markdown extension and is really a syntax helper.

File: test/docfx.Test/docfx.Test.csproj [hunk 8]
```diff
@@ -25,7 +25,7 @@
   </ItemGroup>
 
   <ItemGroup>
-    <ProjectReference Include="../../src/docfx/docfx.csproj" />
+    <ProjectReference Include="..\..\src\docfx\docfx.csproj" />
```
[the remark this answers is not in the document]
Comment (xichan-ms) [reply test/docfx.Test/docfx.Test.csproj:28]: It's changed by vs. I restore it just now.

File: src/docfx/lib/markdown/validation/DocsValidationExtension.cs (hunk 7 above)
Comment (xichan-ms) [reply src/docfx/lib/markdown/validation/DocsValidationExtension.cs:139]: For now, there are no more check. We do this check in order to exclude the Indented CodeBlock which is in Triple Colon.

Comment (yufeih) [reply src/docfx/lib/markdown/validation/DocsValidationExtension.cs]: This check is here to workaround the problem that triple colon block cannot tell apart leaf block vs container block. The fix on triple colon is tracked here: https://ceapex.visualstudio.com/Engineering/_workitems/edit/279439

Comment (xichan-ms) [reply src/docfx/lib/markdown/validation/DocsValidationExtension.cs]: OK.

**Comment by xichan-ms** [comment]
By comparing two files, there are three differences:

(1) Expected. Original error message has the wrong column number, we correct it.
Example, line 134 in https://github.com/MicrosoftDocs/VBA-Docs/edit/master/Library-Reference/Concepts/customize-the-office-fluent-ribbon-by-using-a-managed-com-add-in.md

(2) New found Fenced CodeBlock Invalid Language
Example, line 1354 in https://github.com/MicrosoftDocs/azure-docs-pr/edit/master/articles/cognitive-services/immersive-reader/includes/quickstarts/immersive-reader-client-library-java-android.md

(3) New found Indented CodeBlock
Example, line 148 in https://github.com/MicrosoftDocs/azure-docs-pr/edit/master/articles/azure-functions/durable/durable-functions-create-first-csharp.md

File: src/docfx/lib/markdown/validation/DocsValidationExtension.cs [hunk 9]
```diff
@@ -88,9 +50,108 @@ internal static class DocsValidationExtension
...
                 }
+
+                foreach (var codeBlockItem in codeBlockItemList)
+                {
+                    contentValidator.ValidateCodeBlock(currentFile, codeBlockItem);
+                }
             });
         }
 
+        private static void HeadingValidation(
```
Comment (OsmondJiang) [inline src/docfx/lib/markdown/validation/DocsValidationExtension.cs:61]: BuildHeadingNodes

File: src/docfx/lib/markdown/validation/DocsValidationExtension.cs [hunk 10]
```diff
@@ -88,9 +50,108 @@ internal static class DocsValidationExtension
...
+                    {
+                        inclusionDocumentNodes[context.Document] = inclusionNodes = new List<ContentNode>();
+                    }
+
+                    inclusionNodes.Add(documentNode);
+                }
+            }
+        }
+
+        private static void CodeBlockValidation(
```
Comment (OsmondJiang) [inline src/docfx/lib/markdown/validation/DocsValidationExtension.cs:116]: BuildCodeBlockNodes

File: docs/specs/validations/metadataContentValidations.yml [hunk 11]
```diff
@@ -2461,3 +2461,187 @@ outputs:
...
+    ---
+
+    :::image type="complex" source="image.png" alt-text="triple-colon test":::
+         this is an intented codeblock included in triple colon block
+    :::image-end:::
+         this is an intented codeblock
+outputs:
+  a.json:
+  .errors.log: |
+    {"message_severity":"warning","log_item_type":"user","code":"file-not-found","message":"Invalid file link: 'image.png'.","file":"a.md","line":5,"end_line":5,"column":1,"end_column":1}
```
Comment (XiaofuHuang) [inline docs/specs/validations/metadataContentValidations.yml:2502]: Can you add a file image.png in the test to aviod useless validation result in this test case?

File: src/docfx/validation/ContentValidator.cs [hunk 12]
```diff
@@ -55,6 +55,15 @@ public void ValidateImageLink(Document file, SourceInfo<string> link, string? al
             }
         }
 
+        public void ValidateCodeBlock(Document file, CodeBlockItem codeBlockItem)
```
Comment (OsmondJiang) [inline src/docfx/validation/ContentValidator.cs:58]: bool isIncluded

File: docs/specs/validations/metadataContentValidations.yml (hunk 11 above)
Comment (xichan-ms) [reply docs/specs/validations/metadataContentValidations.yml]: ok

File: src/docfx/lib/markdown/validation/DocsValidationExtension.cs [hunk 13]
```diff
@@ -88,9 +53,93 @@ internal static class DocsValidationExtension
...
+                headingNode.RenderedPlainText = markdownEngine.ToPlainText(headingBlock); // used for validation
+                headingNode.IsVisible = MarkdigUtility.IsVisible(headingBlock);
+
+                documentNode = headingNode;
+            }
+            else if (node is LeafBlock leafBlock)
+            {
+                var contentNode = CreateValidationNode<ContentNode>(context, isCanonicalVersion);
+
+                contentNode.SourceInfo = node.GetSourceInfo();
```
Comment (OsmondJiang) [inline src/docfx/lib/markdown/validation/DocsValidationExtension.cs:90]: put these two into CreateValidationNode

Comment (xichan-ms) [reply src/docfx/lib/markdown/validation/DocsValidationExtension.cs:90]: Got it.
SourceInfo is ok,  but IsVisible belongs to ContentNode.

File: src/docfx/lib/markdown/MarkdigUtility.cs [hunk 14]
```diff
@@ -154,28 +154,29 @@ public static void Visit(this MarkdownObject? obj, MarkdownVisitContext context,
...
+                            target = "pivot";
+                        }
+                        if (!string.IsNullOrEmpty(target))
+                        {
+                            context.ZoneStack.Push(target);
+                        }
+                        foreach (var child in tripleColonBlock)
+                        {
+                            Visit(child, context, action);
+                        }
```
[the remark this answers is not in the document]
Comment (xichan-ms) [reply src/docfx/lib/markdown/MarkdigUtility.cs:175]: Through communication, now we only visit zone triple colon block.

File: src/docfx/lib/markdown/MarkdigUtility.cs [hunk 15]
```diff
@@ -154,28 +154,29 @@ public static void Visit(this MarkdownObject? obj, MarkdownVisitContext context,
                     break;
 
                 case TripleColonBlock tripleColonBlock:
```
Comment (yufeih) [inline src/docfx/lib/markdown/MarkdigUtility.cs:156]:
```suggestion
                case TripleColonBlock tripleColonBlock when tripleColonBlock.Extension is ZoneExtension:
```

File: src/docfx/lib/markdown/validation/DocsValidationExtension.cs [hunk 16]
```diff
@@ -88,9 +53,90 @@ internal static class DocsValidationExtension
...
+
+        private static void BuildCodeBlockNodes(
+            MarkdownObject node,
+            MarkdownVisitContext context,
+            List<(bool IsInclude, CodeBlockItem codeBlockItem)> codeBlockItemList,
+            bool? isCanonicalVersion)
+        {
+            CodeBlockItem? codeBlockItem = null;
+
+            if (node is FencedCodeBlock || node.GetType().Name.Equals(nameof(CodeBlock)))
```
Comment (yufeih) [inline src/docfx/lib/markdown/validation/DocsValidationExtension.cs:117]: consider change `GetType().Name.Equals` check into `is CodeBlock`, or better `switch node` for `FencedCodeBlock` and `CodeBlock`

Comment (xichan-ms) [reply src/docfx/lib/markdown/validation/DocsValidationExtension.cs:117]: OK

File: src/docfx/lib/markdown/MarkdigUtility.cs (hunk 15 above)
Comment (xichan-ms) [reply src/docfx/lib/markdown/MarkdigUtility.cs:156]: OK

**Pull request merged**